geoips/plugins/modules/readers/generic_satpy.py

```diff
@@ -0,0 +1,134 @@
+reader_kwargs argument. The reader_kwargs argument should contain a JSON string
+containing a JSON Object mapping call arguments with their desired values.
+
+In order to map the structure of a Satpy Scene to the Dictionary of Datasets, an
+additional argument named "channel_groups" should be provided via reader_kwargs.
+The argument is a dictionary that groups desired channels by resolution.  This
+dictionary's keys should be strings representing the name of each group and the
+values should be a list of channel names that have the same resolution. The name
+of each group will be used as the keys of the output dictionary and the list of
+channels will be grouped together as a Dataset stored as the associated value to
+the key.  So for example the following channel_groups JSON Object could be used
+to read a selection of visible and IR GOES16 data:
+"channel_groups":{"0.5km":["C02"], "2km":["C13", "C14"]}
+
+While Satpy provides a very large number of readers, a very small number of them
+have been used with this plugin.  This plugin should be considered experimental
+and is not guaranteed to work with all of Satpy's readers.
+"""
+
+import logging
+from typing import Any, Dict, List, Optional, Union
+
+import pyresample.geometry
+import satpy
+import xarray as xr
+
+log = logging.getLogger(__name__)
+
+interface = "readers"
+family = "standard"
+name = "generic_satpy"
+
+
+def call(
+    fnames: List[str],
+    source_name: str,
+    platform_name: str,
+    metadata_only: bool = False,
+    chans: Optional[List[str]] = None,
+    area_def: Optional[pyresample.geometry.AreaDefinition] = None,
+    self_register: bool = False,
+    satpy_reader: str = "abi_L1b",
+    channel_groups: Dict[str, List[str]] = {"0.5km": ["C02"]},
+    roi: float = 3000,
+    filter_parameters: Optional[Dict[str, Any]] = None,
+    satpy_reader_kwargs: Optional[Dict[str, Any]] = None,
+    calibration: Union[List[str], str] = "*",
+    resolution: Union[List[float], float] = "*",
+    polarization: Union[List[str], str] = "*",
+    level: Union[List[float], str] = "*",
+    modifiers: Union[List[str], str] = "*",
+    generate: bool = True,
+    unload: bool = True,
+    load_kwargs: Optional[Dict[str, Any]] = None,
+    force_compute: bool = False,
+) -> Dict[str, xr.DataArray]:
+    if platform_name is None:
+        platform_name = satpy_reader
+
+    scene = satpy.Scene(
+        filenames=fnames,
+        reader=satpy_reader,
+        filter_parameters=filter_parameters,
+        reader_kwargs=satpy_reader_kwargs,
+    )
+
+    all_channels = []
+    for channel_names in channel_groups.values():
+        all_channels.extend(channel_names)
+
+    if load_kwargs is None:
+        load_kwargs = {}
+
+    scene.load(
+        all_channels,
+        calibration=calibration,
+        resolution=resolution,
+        polarization=polarization,
+        level=level,
+        modifiers=modifiers,
+        generate=generate,
```

Comment:
The `generate` kwarg controls whether or not composites are generated or rather *attempted* to be generated. So if `true_color` from Satpy can be made without any interpolation/resampling then we'd want `generate` to be `True` (the default). Unless this reader adds a `Scene.resample` call to do the interpolation on the satpy-side then `generate` should probably always be `True`.

This does bring up the bigger question though: should this reader allow users to load Satpy composites? In the cases where interpolation isn't needed then it would work fine, but in the cases where it is needed then you need a `Scene.resample` call. Otherwise, GeoIPS would need some major hand-holding added to produce all the dependencies for a Satpy composite (`generate=False`) and then recreate the `Scene` in a composite/product/algorithm step and then call `Scene.generate_possible_composites(True)` I think.

Reply:
I don't know if @mindyls or @jsolbrig would disagree, but my thought is that this plugin should probably focus on handling simple use cases.  More advanced use cases, such as loading composites, should probably have their own reader plugin.  So my preference would be to simply remove the `generate` kwarg.

Reply:
To add more context to this, the `generate` kwarg is typically used as an optimization by users who create a lot of RGB composites from only a few input bands. By specifying `generate=False` you could resample, for example, 4 Level 1 bands but produce many RGBs from those 4 inputs. The default, if no resampling/interpolation was needed, would be to resample each of the RGB bands individually assuming they were different from each other when they might not be (ex. R in composite 1 is the same as G in composite 2, but Satpy doesn't know that).